internal/clients/nightfall/nightfall_test.go

```diff
@@ -118,26 +117,93 @@ func (n *nightfallTestSuite) TestScan() {
-		{},
+	mockAPIClient.EXPECT().ScanPayload(gomock.Any(), gomock.AssignableToTypeOf(expectedScanReq)).Return(expectedScanResponse, nil, nil)
+
+	resp, err := client.Scan(context.Background(), githublogger.NewDefaultGithubLogger(), items)
+	n.NoError(err, "Received error from Scan")
+	n.Equal(expectedScanResponse, resp, "Received incorrect response from Scan")
+}
+
+func (n *nightfallTestSuite) TestScanRetries() {
+	ctrl := gomock.NewController(n.T())
+	defer ctrl.Finish()
+	cc := nightfallAPI.CREDIT_CARD_NUMBER
+	phone := nightfallAPI.PHONE_NUMBER
+	detectors := []*nightfallAPI.Detector{&cc, &phone}
+
+	items := []string{
+		"this is a string",
+		fmt.Sprintf("this has a credit card number %s", exampleCreditCardNumber),
+		"tom cruise!!!!!!",
+	}
+
+	expectedScanReq := createScanReq(detectors, items)
+	expectedTooManyRequestsHTTPResponse := &http.Response{StatusCode: http.StatusTooManyRequests}
+
+	tests := []struct {
+		haveSuccess        bool
+		haveNumReqAttempts int
+		wantResponse       [][]nightfallAPI.ScanResponse
+		desc               string
+	}{
 		{
-			{
-				Fragment: exampleCreditCardNumber,
-				Detector: string(nightfallAPI.CREDIT_CARD_NUMBER),
-				Confidence: nightfallAPI.ScanResponseConfidence{
-					Bucket: string(nightfallAPI.LIKELY),
-				},
-			},
+			haveSuccess:        true,
+			haveNumReqAttempts: 1,
+			wantResponse:       expectedScanResponse,
+			desc:               "success on first scan request",
+		},
+		{
+			haveSuccess:        true,
+			haveNumReqAttempts: 3,
+			wantResponse:       expectedScanResponse,
+			desc:               "success on third scan request attempt",
+		},
+		{
+			haveSuccess:        false,
+			haveNumReqAttempts: 5,
+			wantResponse:       [][]nightfallAPI.ScanResponse(nil),
+			desc:               "failed after max retries",
 		},
-		{},
 	}
+	for _, tt := range tests {
+		mockAPIClient := nightfallapi_mock.NewNightfallAPI(ctrl)
 
-	mockAPIClient.EXPECT().ScanAPI().Return(mockScanAPI)
-	mockScanAPI.EXPECT().ScanPayload(gomock.Any(), gomock.AssignableToTypeOf(expectedScanReq)).Return(expectedScanResp, nil, nil)
+		client := nightfall.Client{
+			APIClient:         mockAPIClient,
+			Detectors:         detectors,
+			InitialRetryDelay: time.Millisecond,
+		}
 
-	resp, err := client.Scan(context.Background(), githublogger.NewDefaultGithubLogger(), items)
-	n.NoError(err, "Received error from Scan")
-	n.Equal(expectedScanResp, resp, "Received incorrect response from Scan")
+		numRetries := tt.haveNumReqAttempts
+		if tt.haveSuccess {
+			numRetries--
+		}
+		for i := 0; i < numRetries; i++ {
+			mockAPIClient.EXPECT().ScanPayload(gomock.Any(), gomock.AssignableToTypeOf(expectedScanReq)).
+				DoAndReturn(
+					func(
+						ctx context.Context,
+						scanReq nightfallAPI.ScanRequest,
+					) ([][]nightfallAPI.ScanResponse, *http.Response, error) {
+						return [][]nightfallAPI.ScanResponse{nil}, expectedTooManyRequestsHTTPResponse, nightfall.ErrMaxScanRetries
```

Comment:
nit - do we need to use `DoAndReturn` or can we just use `Return`, just trying to simplify this